cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -107,6 +111,322 @@ static void _sc_cgroup_v1_attach_pid(sc_device_cgroup *self, pid_t pid) {
+        /* the v1 implementation blocks all devices by default and then adds
+         * each assigned one individually, however for v2 there's no way to drop
+         * all the contents of the map*/
+
+        /* first collect all keys in the map */
+        sc_cgroup_v2_device_key *existing_keys = calloc(max_entries, sizeof(sc_cgroup_v2_device_key));
+        /* 'current' key is zeroed, such that no entry can match it and thus
+         * we'll iterate over the keys from the beginning */
+        sc_cgroup_v2_device_key key = {0};
+        size_t existing_count = 0;
+        while (true) {
+            sc_cgroup_v2_device_key next = {0};
+            if (existing_count >= max_entries) {
+                die("too many elements in the map");
+            }
+            if (existing_count > 0) {
+                /* grab the previous key */
+                key = existing_keys[existing_count - 1];
+            }
+            int ret = bpf_map_get_next_key(devmap_fd, &key, &next);
+            if (ret == -1) {
+                if (errno != ENOENT) {
+                    die("cannot lookup existing device map keys");
+                }
+                /* we are done */
+                break;
+            }
+            existing_keys[existing_count] = next;
+            existing_count++;
+        }
+        debug("found %zu existing entries in devices map", existing_count);
+        if (existing_count > 0) {
+#if 0
+            /* XXX: we should be doing a batch delete of elements, however:
+             * - on Arch with 5.13 kernel I'm getting EINVAL
+             * - the linux/bpf.h header present during build on 16.04 does not
+             *     support batch operations
+             */
+            if (bpf_map_delete_batch(devmap_fd, existing_keys, existing_count) < 0) {
+                 die("cannot dump all elements from devices map");
+            }
+#endif
+            for (size_t i = 0; i < existing_count; i++) {
+                sc_cgroup_v2_device_key key = existing_keys[i];
+                debug("delete key for %c %d:%d", key.type, key.major, key.minor);
+                if (bpf_map_delete_elem(devmap_fd, &key) < 0) {
+                    die("cannot delete device map entry for %c %d:%d", key.type, key.major, key.minor);
+                }
+            }
+        }
+    }
+
+    if (flags != SC_DEVICE_CGROUP_FROM_EXISTING) {
+        int prog_fd = load_devcgroup_prog(devmap_fd);
+
+        if (bpf_prog_attach(BPF_CGROUP_DEVICE, cgroup_fd, prog_fd) < 0) {
+            die("cannot attach cgroup program");
+        }
+    }
+
+    self->v2.devmap_fd = devmap_fd;
+    self->v2.cgroup_fd = cgroup_fd;
+
+    return 0;
+}
+
+static void _sc_cgroup_v2_close_bpf(sc_device_cgroup *self) {
+    sc_cleanup_string(&self->v2.tag);
+    /* the map is pinned to a per-snap-application file and referenced by the
+     * program */
+    sc_cleanup_close(&self->v2.devmap_fd);
+    sc_cleanup_close(&self->v2.cgroup_fd);
```

Comment:
Does the map path need to be `unlink()`'d as well? ie. 

```C
char path[PATH_MAX] = {0};
sc_must_snprintf(path, sizeof path, "/sys/fs/bpf/snap/%s", self->v2.tag);
unlink(path);
```

Or does closing the fd do this automatically?

Reply:
We do not unlink the map at all. It remains there, so that snap-device-helper can access it and do the update if the devices change. In this sense it's similar to what we do for cgroup v1, the group stays behind and it only gets updated when a new application starts or udev pokes snap-device-helper.

Reply:
Ah ok, that makes sense.